core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -90,30 +89,252 @@
   /** Whether to cache blocks in this file into Alluxio. */
   private final boolean mShouldCache;
 
-  // The following 3 fields must be kept in sync. They are only updated in updateStreams together.
+  // mCurrentBlockInStream, and mCurrentCacheStream should always be in sync.
+  //
+  // When mPositionState is updated, the above two values are not updated until calling
+  // updateStreams, so they might be out of sync with the current position states.
+  //
+  // But after calling updateStreams, they should be in sync with mPositionState, for example,
+  // mPosition.getBlockId should equal mCurrentBlockInStream.getId(),
+  // mPositionState.getBlockStartPos should be the start position of mCurrentBlockInStream, and
+  // mPositionState.getNextBlockStartPos should be the exclusive end position of
+  // mCurrentBlockInStream.
+  /** Position states. */
+  protected PositionState mPositionState;
   /** Current block in stream backing this stream. */
   protected BlockInStream mCurrentBlockInStream;
   /**
    * Current block out stream writing the data into the local worker. This is only used when the in
    * stream reads from a remote worker.
    */
   protected BlockOutStream mCurrentCacheStream;
-  /** The blockId used in the block streams. */
-  private long mStreamBlockId;
 
   /** The read buffer in file seek. This is used in {@link #readCurrentBlockToEnd()}. */
   private final byte[] mSeekBuffer;
 
   /**
+   * States related to position that need to be in sync.
+   */
+  protected class PositionState {
+    /** Block ID used when position reaches EOF, should be < 0. */
+    public static final long EOF_BLOCK_ID = -1;
+
+    // These are not states related to position, but are constants needed for calculating and
+    // maintaining the invariants in the position related states.
+    /** File length. */
+    private final long mFileLength;
+    /** Block IDs. */
+    private final List<Long> mBlockIds;
+    /** Block size in bytes. */
+    private final long mBlockSize;
+
+    /**
+     * Current position of the file.
+     * Initially, it is 0.
+     * When EOF is reached, it is the file length.
+     */
+    private long mPos;
+    /** The remaining bytes of the file. */
+    private long mRemaining;
+
+    // States derived from mPos.
+    /**
+     * Index of the block where the current file position belongs to.
+     * Initially, it is 0.
+     * When EOF is reached, it is the size of mStatus.getBlockIds.
+     */
+    private int mBlockIndex;
+    /**
+     * ID of the block where the current file position belongs to.
+     * Initially, it is the first block ID.
+     * When EOF is reached, it is {@link #EOF_BLOCK_ID}.
+     */
+    private long mBlockId;
+    /**
+     * Start position of the block where the current file position belongs to.
+     * Initially, it is 0.
+     * When EOF is reached, it is the file length.
```

Comment:
Why do we need to set this to file length when EOF is reached? From its usage, it seems okay to keep it as the start position of the last block when EOF is reached.

Reply:
otherwise, when EOF is reached, `isInCurrentBlock(fileLength-1)` would return `true`.

Reply:
I see. We also use this to indicate if EOF is reached.